src/opencensus/proto/trace/v1/BUILD.bazel

```diff
@@ -14,66 +14,36 @@
 
 package(default_visibility = ["//visibility:public"])
 
-load("@org_pubref_rules_protobuf//go:rules.bzl", "go_proto_library")
+load("@io_bazel_rules_go//proto:def.bzl", "go_proto_library")
 
 proto_library(
     name = "trace_proto",
-    srcs = ["trace.proto"],
+    srcs = [
+        "trace.proto",
+        "trace_config.proto"
+    ],
     deps = [
         "@com_google_protobuf//:timestamp_proto",
         "@com_google_protobuf//:wrappers_proto",
     ],
 )
 
-proto_library(
```

Comment:
Why do you do this? It should work with having them in different entries.

Reply:
When the trace and trace_config rules are separate there is an error building the `//opencensus/proto/agent/trace/v1:trace_service_proto_go` rule. See: https://travis-ci.org/census-instrumentation/opencensus-proto/builds/463521132?utm_source=github_status&utm_medium=notification

I suspect it's because there are two `go_proto_library` rules for `trace_proto_go` and `trace_config_proto_go`, but they both share the same `importpath`. Combining them into a single rule allows the build to pass.

Am I missing something here? Any thoughts on how to do this better?

Reply:
I think I need to create a separate Go proto build rule with the `paths=source_relative` flag and then pass it to the `compiler` option of `go_proto_library`.

I'll plan to work on that some time soon.

Reply:
Slight correction, the relevant failing Travis build is https://travis-ci.org/census-instrumentation/opencensus-proto/jobs/463529698

It says `undefined: "github.com/census-instrumentation/opencensus-proto/gen-go/trace/v1".TraceConfig`

Reply:
I think it's better to keep `trace.proto` and `trace_config.proto` separate since in most cases users only need `trace.proto`.

cc @yancl for suggestions on go_proto_library.

Reply:
I thinks maybe this is a bug of **io_bazel_rules_go**(which filtered out the trace_config.proto).

I made the following diagnose:
* `bazel build -s //...` to output the sub commands with the **-s** options
* it outputs something like this:
```
(cd /home/vagrant/.cache/bazel/_bazel_vagrant/39f7101e26fcabfa22265379537b720c/execroot/opencensus_proto && \
  exec env - \
    CGO_ENABLED=1 \
    GOARCH=amd64 \
    GOOS=linux \
    GOROOT=external/go_sdk \
    GOROOT_FINAL=GOROOT \
    PATH=/usr/bin:/bin \
  bazel-out/host/bin/external/io_bazel_rules_go/go/tools/builders/linux_amd64_stripped/compile **-v** -sdk external/go_sdk ...
```
* I add a **-v** args to the above command to output the real `go tool compile` commands like the following:
```
CGO_ENABLED=1 \
GOARCH=amd64 \
GOOS=linux \
GOROOT=external/go_sdk \
GOROOT_FINAL=GOROOT \
PATH=/usr/bin:/bin \
external/go_sdk/pkg/tool/linux_amd64/compile -importcfg /home/vagrant/.cache/bazel/_bazel_vagrant/39f7101e26fcabfa22265379537b720c/execroot/opencensus_proto/bazel-out/k8-fastbuild/bin/opencensus/proto/agent/trace/v1/linux_amd64_stripped/trace_service_proto_go%/github.com/census-instrumentation/opencensus-proto/gen-go/agent/trace/importcfg436570685 -pack -o /home/vagrant/.cache/bazel/_bazel_vagrant/39f7101e26fcabfa22265379537b720c/execroot/opencensus_proto/bazel-out/k8-fastbuild/bin/opencensus/proto/agent/trace/v1/linux_amd64_stripped/trace_service_proto_go%/github.com/census-instrumentation/opencensus-proto/gen-go/agent/trace/v1.a -trimpath /home/vagrant/.cache/bazel/_bazel_vagrant/39f7101e26fcabfa22265379537b720c/execroot/opencensus_proto -p github.com/census-instrumentation/opencensus-proto/gen-go/agent/trace/v1 -- /home/vagrant/.cache/bazel/_bazel_vagrant/39f7101e26fcabfa22265379537b720c/execroot/opencensus_proto/bazel-out/k8-fastbuild/bin/opencensus/proto/agent/trace/v1/linux_amd64_stripped/trace_service_proto_go%/github.com/census-instrumentation/opencensus-proto/gen-go/agent/trace/v1/trace_service.pb.go
```
* As you see there is **-importcfg** args, then I find the **trace/v1.a** in the depends file:
```
$ cat /home/vagrant/.cache/bazel/_bazel_vagrant/39f7101e26fcabfa22265379537b720c/execroot/opencensus_proto/bazel-out/k8-fastbuild/bin/opencensus/proto/agent/trace/v1/linux_amd64_stripped/trace_service_proto_go%/github.com/census-instrumentation/opencensus-proto/gen-go/agent/trace/importcfg722808585
packagefile fmt=/home/vagrant/.cache/bazel/_bazel_vagrant/39f7101e26fcabfa22265379537b720c/execroot/opencensus_proto/external/go_sdk/pkg/linux_amd64/fmt.a
packagefile math=/home/vagrant/.cache/bazel/_bazel_vagrant/39f7101e26fcabfa22265379537b720c/execroot/opencensus_proto/external/go_sdk/pkg/linux_amd64/math.a
...

packagefile github.com/census-instrumentation/opencensus-proto/gen-go/trace/v1=/home/vagrant/.cache/bazel/_bazel_vagrant/39f7101e26fcabfa22265379537b720c/execroot/opencensus_proto/bazel-out/k8-fastbuild/bin/opencensus/proto/trace/v1/linux_amd64_stripped/trace_proto_go%/github.com/census-instrumentation/opencensus-proto/gen-go/trace/v1.a
...
```

* Then I dump the symbols of the **v1.a** using golang **nm** and can't find **TraceConfig**
```
$ ./bazel-src/external/go_sdk/pkg/tool/linux_amd64/nm /home/vagrant/.cache/bazel/_bazel_vagrant/39f7101e26fcabfa22265379537b720c/execroot/opencensus_proto/bazel-out/k8-fastbuild/bin/opencensus/proto/trace/v1/linux_amd64_stripped/trace_proto_go%/github.com/census-instrumentation/opencensus-proto/gen-go/trace/v1.a | grep TraceConfig
vagrant@vbox:~/3rd/opencensus-proto/src$
```

* But with the correct **v1.a**  build by @draffensperger previous branch(fix-bazel-build2) I can find the **TraceConfig**
```
$ ./bazel-src/external/go_sdk/pkg/tool/linux_amd64/nm ~/3rd/x/opencensus-proto/src/bazel-bin/opencensus/proto/trace/v1/linux_amd64_stripped/trace_proto_go%/github.com/census-instrumentation/opencensus-proto/gen-go/trace/v1.a | grep TraceConfig | head -n 10
   35a87 T %22%22.(*TraceConfig).Descriptor
   35faf T %22%22.(*TraceConfig).GetConstantSampler
```

So maybe I need some more debug for why the trace_config.proto filtered out if I was right:)

Reply:
Thanks for the investigation here. I posted an issue for this as well: https://github.com/bazelbuild/rules_go/issues/1841

One workaround would be to create a new separate rule `proto_library` called `trace_and_config` that combines the trace.proto and trace_config.proto, and then depend on that one for the trace agent service. We could comment to explain this issue.

Reply:
Based on feedback on that issue (https://github.com/bazelbuild/rules_go/issues/1841#issuecomment-444665833), it seems like the best approach is to use an intermediate `go_library` rule that embeds the trace and trace config `go_proto_library`'s, and then depend on that in the trace agent service. I did that and added a comment with a link to the GitHub issue.

Reply:
And thanks @yancl for the detailed investigation!